abba/json/src/Parse.cpp

```diff
@@ -0,0 +1,480 @@
+#include "Parse.hpp"
+
+Parse::Parse(std::string filename) : json(filename)
+{
+
+}
+
+void Parse::parse()
+{
+	_file = &json.getFile();
+	std::cout << "i'm here" << std::endl;
+	getToken();
+	Json *root = new Json();
+	root->type = Json::OBJECT;
+	root = parseObject(root);
+	std::cout << "DONE PARSING!!!!!!!!!!!!" << std::endl;
+}
+
+char Parse::getWithoutWhiteSpace()
+{
+	std::cout << "getWithoutWhiteSpace function called" << std::endl;
+	char	c = ' ';
+	while (c == ' ' || c == '\t' || c == '\n' || \
+			c == '\v' || c == '\f' || c == '\r')
+	{
+		_file->get(c);
+		if (!_file)
+			return c;
+	}
+	return c;
+}
+
+Parse::Type Parse::getToken()
+{
+	char		c;
+	Type	type;
+
+//	std::cout << "getToken function called" << std::endl;
+	if (_file->eof())
+	{
+		std::cout << "finish reading" << std::endl;
+	}
+	_prevPos = _file->tellg();
+	c = getWithoutWhiteSpace();
+	if (c == '"')
+	{
+		type.type = STRING;
+		_file->get(c);
+		while (c != '\n' && c != '"')
+		{
+			type.value += c;
+			_file->get(c);
+		}
+	}
+	else if (c == '[')
+	{
+//		std::cout << "array open" << std::endl;
+		type.type = ARRAY_OPEN;
+	}
+	else if (c == ']')
+	{
+//		std::cout << "array close" << std::endl;
+		type.type = ARRAY_CLOSE;
+	}
+	else if (c == '{')
+	{
+//		std::cout << "curly open" << std::endl;
+		type.type = CURLY_OPEN;
+	}
+	else if (c == '}')
+		type.type = CURLY_CLOSE;
+	else if (c == ':')
+		type.type = COLON;
+	else if (c == ',')
+		type.type = COMMA;
+	else if (c >= '0' && c <= '9')
+	{
+		type.type = NUMBER;
+		while (c >= '0' && c <= '9')
+		{
+			_prevPos = _file->tellg();
+			type.value += c;
+			_file->get(c);
+		}
+		_file->seekg(_prevPos);
+	}
+	else if (c == 'n')
+	{
+		type.type = NULL_TYPE;
+		while (isalpha(c))
```

Comment:
Since there are a lot of loops you could implement a generic function that takes a set of characters as long as some function returns true. `value = extract_characters(isalpha)` and for more complex conditionals you can pass a lambda (anonymous) function: `[] (char c) { return c != '"'; }`

Reply:
I'll check about anonymous lambda, it seems interesting!

Reply:
it's essentially a function pointer, you could argue that lambda's are sometimes less clear because they aren't named